lib/srv/discovery/status.go

```diff
@@ -0,0 +1,167 @@
+/*
+ * Teleport
+ * Copyright (C) 2024  Gravitational, Inc.
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+package discovery
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	"github.com/gravitational/trace"
+
+	discoveryconfigv1 "github.com/gravitational/teleport/api/gen/proto/go/teleport/discoveryconfig/v1"
+	"github.com/gravitational/teleport/api/types/discoveryconfig"
+	aws_sync "github.com/gravitational/teleport/lib/srv/discovery/fetchers/aws-sync"
+)
+
+// updateDiscoveryConfigStatus updates the DiscoveryConfig Status field with the current in-memory status.
+// The status will be updated with the following matchers:
+// - AWS Sync (TAG) status
+func (s *Server) updateDiscoveryConfigStatus(discoveryConfigName string) {
+	discoveryConfigStatus := discoveryconfig.Status{
+		State:        discoveryconfigv1.DiscoveryConfigState_DISCOVERY_CONFIG_STATE_SYNCING.String(),
+		LastSyncTime: s.clock.Now(),
+	}
+
+	// Merge AWS Sync (TAG) status
+	discoveryConfigStatus = s.awsSyncStatus.mergeIntoGlobalStatus(discoveryConfigName, discoveryConfigStatus)
+
+	ctx, cancel := context.WithTimeout(s.ctx, 5*time.Second)
+	defer cancel()
+
+	_, err := s.AccessPoint.UpdateDiscoveryConfigStatus(ctx, discoveryConfigName, discoveryConfigStatus)
+	switch {
+	case trace.IsNotImplemented(err):
+		s.Log.Warn("UpdateDiscoveryConfigStatus method is not implemented in Auth Server. Please upgrade it to a recent version.")
+	case err != nil:
+		s.Log.WithError(err).WithField("discovery_config_name", discoveryConfigName).Info("Error updating discovery config status")
+	}
+}
+
+// awsSyncStatus contains all the status for aws_sync Fetchers grouped by DiscoveryConfig.
+type awsSyncStatus struct {
+	mu sync.RWMutex
+	// awsSyncStatus maps the DiscoveryConfig name to a aws_sync result.
+	// Each DiscoveryConfig might have multiple `aws_sync` matchers.
+	awsSyncStatus map[string][]awsSyncResult
+}
```

Comment:
nit: A field `awsSyncStatus` in a type with the same name is bit confusing.

Reply:
Agreed. Renamed to `awsSyncResults`